python/tritonserver/_api/_datautils.py

```diff
@@ -0,0 +1,327 @@
+
+        if hasattr(self, "set_attributes"):
+            self._allocator.set_buffer_attributes_function(self.set_buffer_attributes)
+        return self._allocator
+
+
+try:
+    import cupy
+
+    class CupyAllocator(MemoryAllocator):
+        def __init__(self):
+            pass
+
+        def start(self, allocator, user_object):
+            pass
+
+        def release(
+            self,
+            allocator,
+            buffer_,
+            buffer_user_object,
+            byte_size,
+            memory_type,
+            memory_type_id,
+        ):
+            pass
+
+        def allocate(
+            self,
+            allocator,
+            tensor_name,
+            byte_size,
+            memory_type,
+            memory_type_id,
+            user_object,
+        ):
+            _buffer = cupy.empty(byte_size, cupy.byte)
+
+            dlpack_object = DLPackObject(_buffer)
+
+            return (
+                dlpack_object.data_ptr,
+                _buffer,
+                dlpack_object.memory_type,
+                dlpack_object.memory_type_id,
+            )
+
+except Exception as e:
+    pass
+
+
+class NumpyAllocator(MemoryAllocator):
+    def __init__(self):
+        pass
+
+    def start(self, allocator, user_object):
+        pass
+
+    def allocate(
+        self,
+        allocator,
+        tensor_name,
+        byte_size,
+        memory_type,
+        memory_type_id,
+        user_object,
+    ):
+        _buffer = numpy.empty(byte_size, numpy.byte)
+        return (
+            _buffer.ctypes.data,
+            _buffer,
+            triton_bindings.TRITONSERVER_MemoryType.CPU,
+            0,
+        )
+
+    def release(
+        self,
+        allocator,
+        buffer_,
+        buffer_user_object,
+        byte_size,
+        memory_type,
+        memory_type_id,
+    ):
+        pass
+
+
+class DLPackObject:
+    def __init__(self, value):
+        self._capsule = _dlpack.get_dlpack_capsule(value)
+        self._tensor = _dlpack.get_managed_tensor(self._capsule).dl_tensor
+
+    @property
+    def byte_size(self):
+        return _dlpack.get_byte_size(
+            self._tensor.dtype, self._tensor.ndim, self._tensor.shape
+        )
+
+    @property
+    def memory_type(self):
+        return DLPACK_DEVICE_TYPE_TO_TRITON_MEMORY_TYPE[self._tensor.device.device_type]
+
+    @property
+    def memory_type_id(self):
+        return self._tensor.device.device_id
+
+    @property
+    def shape(self):
+        return [self._tensor.shape[i] for i in range(self._tensor.ndim)]
+
+    @property
+    def triton_data_type(self):
+        return DLPACK_TO_TRITON_DTYPE[self.data_type]
+
+    @property
+    def data_type(self):
+        return (self._tensor.dtype.type_code, self._tensor.dtype.bits)
+
+    @property
+    def data_ptr(self):
+        return self._tensor.data + self._tensor.byte_offset
+
+
+@dataclass
+class MemoryBuffer:
```

Comment:
perhaps rename to Tensor?

Reply:
wanted to avoid considering it a tensor specifically because I felt the term gets overloaded - but can revisit.

Reply:
Will rename to Tensor